compiler/rustc_middle/src/mir/interpret/value.rs

```diff
@@ -38,27 +35,38 @@ pub enum ConstValue<'tcx> {
     /// Only used for ZSTs.
     ZeroSized,
 
-    /// Used only for `&[u8]` and `&str`
-    Slice { data: ConstAllocation<'tcx>, start: usize, end: usize },
-
-    /// A value not represented/representable by `Scalar` or `Slice`
-    ByRef {
-        /// The backing memory of the value, may contain more memory than needed for just the value
-        /// in order to share `ConstAllocation`s between values
-        alloc: ConstAllocation<'tcx>,
+    /// Used only for `&[u8]` and `&str`.
+    ///
+    /// This is worth the optimization since Rust has literals of that type.
+    /// See `Indirect` for why we are using an `AllocId`.
+    /// We allow `None` for empty slices; in that case `start == end` and `start` is the absolute
+    /// address of the slice.
+    Slice { alloc_id: Option<AllocId>, start: usize, end: usize },
```

Comment:
I think we can avoid the `None` case... literals always get an `AllocId`, so this can only happen as the result of `ConstEval`, and then we could use `ConstValue::Indirect` since we already have an allocation that stores the evaluation result. OTOH, with `None` we can guarantee that all `&str`/`&[u8]` are `ConstValue::Slice`; do we rely on that somewhere -- e.g. in pattern matching?

Reply:
Ah, but then we'd have a really subtle invariant saying that if the value fed to `op_to_const` is an immediate slice reference then it must point to an actual allocation. Though maybe that's worth it for simplifying literally every consumer of `ConstValue::Slice` to avoid the `Option` here.

Reply:
Yea, simpler is probably better